### PR TITLE
Backend adapters - proof of concept

### PR DESCRIPTION
**Pull request by b4rtaz**
---
name: Backend adapters - proof of concept
---

### CODE REVIEW

File: packages/client/evmAuth/src/MetaMaskEvmConnector.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { Web3Provider } from '@ethersproject/providers';
+import { AuthClientError, AuthClientErrorCode, WalletDetails } from '@moralisweb3/client-auth-utils';
+import { EvmConnection, EvmConnector } from './EvmConnector';
+import detectEthereumProvider from '@metamask/detect-provider';
+
+export class MetaMaskEvmConnector implements EvmConnector {
+  public readonly name = 'default';
```
Comment (ErnoW) [inline packages/client/evmAuth/src/MetaMaskEvmConnector.ts:7]: Let's rename this to metamask

File: demos/webpack/src/main.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
+import { MoralisClient } from '@moralisweb3/client';
+import { FrontEndOnlyBackendAdapter } from '@moralisweb3/client-backend-adapter-frontend-only';
+import { Connection, clusterApiUrl, PublicKey } from '@solana/web3.js';
+import { WalletConnectEvmConnector } from '@moralisweb3/client-connector-wallet-connect';
+
+async function reloadCurrentUser() {
+  const userAddress = document.getElementById('userAddress') as HTMLElement;
+  const userBalance = document.getElementById('userBalance') as HTMLElement;
+
+  const users = await Promise.all([MoralisClient.EvmAuth.tryGetUser(), MoralisClient.SolAuth.tryGetUser()]);
```
Comment (ErnoW) [inline demos/webpack/src/main.ts:10]: Probably a good utility in the future is to have a `tryGetUser` on `MoralisClient`, that calls this function for all network types, like you do here

File: packages/common/core/src/Modules/Module.ts [hunk 3]
```diff
@@ -1,52 +1,23 @@
...
-  public abstract start(): void | Promise<void>;
-
-  /**
-   * Any cleanup that needs to be done for removing this module.
-   * It also should remove the module via `this.core.modules.remove(this.name)`
-   */
-  public cleanUp() {
-    this.core.modules.remove(this.name);
-  }
+export interface Module {
```
Comment (ErnoW) [inline packages/common/core/src/Modules/Module.ts:13]: Nice cleanup, as we don't need the eventlisteners by default anymore 🎉

File: packages/client/authUtils/package.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    "lib/*"
+  ],
+  "scripts": {
+    "test": "yarn jest --runInBand --detectOpenHandles --forceExit --ci",
+    "test:watch": "yarn run test --watch",
+    "lint": "eslint . --ext .js,.ts,.tsx,jsx",
+    "clean": "rm -rf lib && rm -rf tsconfig.tsbuildinfo && rm -rf ./node_modules/.cache/nx",
+    "build": "tsc",
+    "dev": "tsc --watch",
+    "gen:client": "node ../apiUtils/scripts/generate-client.js ./src/generated/ClientSolApi.ts @moralisweb3/common-sol-utils"
```
Comment (ErnoW) [inline packages/client/authUtils/package.json:19]: Does not work anymore

File: packages/client/authUtils/src/AuthClientError.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+export enum AuthClientErrorCode {
+  GENERIC = 'CA0000',
+  ALREADY_AUTHENTICATED = 'CA0001',
+  NOT_AUTHENTICATED = 'CA0002',
+  ALREADY_CONNECTED = 'CA0003',
+  NOT_CONNECTED = 'CA0004',
+}
+
+// TODO: we don't use errors from the @moralisweb3/core package, because we would have a collision with the MoralisAuthError type.
```
Comment (ErnoW) [inline packages/client/authUtils/src/AuthClientError.ts:9]: Seems like a good reason to prioritise the structure of error codes. Ideally we want them to be in the specified packages. 👍

File: packages/client/authUtils/src/Connector.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  provider: WalletProvider;
+  readWallet(): Promise<WalletDetails>;
+  signMessage(message: string): Promise<string>;
+  disconnect(): Promise<void>;
+}
+
+export interface WalletDetails {
+  address: string;
+  chain?: string;
+  network?: string;
```
Comment (ErnoW) [inline packages/client/authUtils/src/Connector.ts:17]: Lets rename this networkType, as we had issues in the past where users confuse network with chain.

File: packages/client/backendAdapter/frontendOnly/src/FrontEndOnlyAuth.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        networkType,
+        address: session.address,
+        // TODO: here we should call the Moralis API and receive the profileId.
+        profileId: `TODO_DO_NOT_USE_THIS_VALUE_${session.address}_TODO`,
+      };
+    }
+    return null;
+  }
+
+  public async getMessageToSign(request: GetMessageToSignRequest): Promise<GetMessageToSignResponse> {
```
Comment (ErnoW) [inline packages/client/backendAdapter/frontendOnly/src/FrontEndOnlyAuth.ts:31]: This only signs the message without using our auth service, right?
We already have a task for implementing auth with frontendOnly adapter, when our auth serivice supports this

File: packages/client/backendAdapter/utils/src/NetworkTypeResolver.test.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { NetworkTypeResolver } from './NetworkTypeResolver';
```
Comment (ErnoW) [inline packages/client/backendAdapter/utils/src/NetworkTypeResolver.test.ts:1]: Can you add test coverage reporting to our CI? 
in `main.yml` > `Test coverage comment`

The same holds for the other packages that have been added that include tests

**Review by ErnoW — CHANGES_REQUESTED** [review]

File: demos/webpack/src/main.ts (hunk 2 above)
Comment (b4rtaz) [reply demos/webpack/src/main.ts:10]: Nice idea! I'll add this.

File: packages/client/authUtils/src/Connector.ts (hunk 6 above)
Comment (b4rtaz) [reply packages/client/authUtils/src/Connector.ts:17]: Here `network` doesn't mean the network type. This is a Solana network. I'll rename it to `solNetwork` and add some comment.

File: packages/client/backendAdapter/frontendOnly/src/FrontEndOnlyAuth.ts (hunk 7 above)
Comment (b4rtaz) [reply packages/client/backendAdapter/frontendOnly/src/FrontEndOnlyAuth.ts:31]: > This only signs the message without using our auth service, right?

Yes.

When the API will be ready we can refactor this file to support the changes.

**Review by ErnoW — APPROVED** [review]

**Pull request merged**